[RISCV][TTI] Properly model odd vector sized LD/ST operations

preames opened the pull request:
The motivation for this change is the costing of a LD or ST with nearly power of 2 vectors (e.g. <3 x i32> or <7 x i32>) on V.  There's an experimental option in SLP to allow emitting these if the cost model says they're profitable.  This really helps with e.g. RGB vectors.  

Our actual lowering for these depends on whether a wider container type is known available.  If so, we use a vle or vse on the wider type with a restricted VL.  If not, we split until a legal type is found, and then apply the vle/vse on the sub-pieces.  

This change is intentionally restricted to only the case where promotion (widening w/VL predication) is involved.  We appear to have at least one bug in our splitting lowering (see discussion on review), and to avoid exposing this more widely, I chose to not adjust costs for the splitting case.  The current splitting costing assumes scalarization (which is not true of the actual lowering), but that has the effect of biasing vectorization away from such cases strongly.

For the widening case, the true cost scales with the next largest legal type.  The default implementation assumes that such a type is scalarized.  Changing that brings our cost in line with our actual lowering decision.  Note that since scalarization is not possible for scalable types, the prior costing falsely returned Invalid for that case.

preames committed:
[RISCV][TTI] Properly model odd vector sized LD/ST operations

Our actual lowering for these uses RVV instructions if a wider
container type is available, and splits if not.  The splitting
case was reasonable handled costing wise, but not the widening.
For that case, the true cost scales with the next largest legal
type.  The default implementation assumes that such a type is
scalarized.  For scalable types which can be promoted (which
seems to only happen on zve32x), this would result in an illegal
cost being returned spuriously.

preames commented:
A couple bits of context here.

The goal here is to get simple cases with -slp-vectorize-non-power-of-2 working.  We're probably got other places in the cost model which need fixed here, but I'm starting with a simple one.

This code could be better integrated with the default costing.  I do want to take a look at that, but it's slightly more involved than I wanted to couple into the -slp-vectorize-non-power-of-2 work.

preames commented:
Hold off on reviewing this, I think I may have pushed a bad commit.

preames commented:
> Hold off on reviewing this, I think I may have pushed a bad commit.

Corrected version pushed.  Sorry, I appear to have screwed up a rebase/squash for the first version posted.

lukel97 commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -1390,14 +1390,34 @@ InstructionCost RISCVTTIImpl::getMemoryOpCost(unsigned Opcode, Type *Src,
+      auto *SrcVT = cast<VectorType>(Src);
+      TypeSize SrcEltSize = DL.getTypeStoreSizeInBits(SrcVT->getElementType());
+      TypeSize LegalEltSize = LT.second.getVectorElementType().getSizeInBits();
+      if (SrcEltSize == LegalEltSize)
```
I presume the case this is trying to catch is <n x i64> on zve32x, but wouldn't that be expanded not promoted?

preames replied:
The primary motivation is <3 x i32> on V.  The scalable vector thing on zve32x was merely a side finding.  And no, that would be a single vle on the wider legal type.  Here's a standalone test which shows that:

```
; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
; RUN: llc -mtriple=riscv64 -mattr=+zve32x < %s | FileCheck %s

define void @test(ptr %a, ptr %b) nounwind vscale_range(4,1024) {
; CHECK-LABEL: test:
; CHECK:       # %bb.0:
; CHECK-NEXT:    csrr a2, vlenb
; CHECK-NEXT:    srli a2, a2, 3
; CHECK-NEXT:    vsetvli zero, a2, e32, m1, ta, ma
; CHECK-NEXT:    vle32.v v8, (a0)
; CHECK-NEXT:    vse32.v v8, (a1)
; CHECK-NEXT:    ret
  %LD = call <vscale x 1 x i32> @llvm.masked.load.nxv1i32.p0(ptr %a, i32 4, <vscale x 1 x i1> splat(i1 true), <vscale x 1 x i32> poison)
  store <vscale x 1 x i32> %LD, ptr %b
  ret void
}

```

lukel97 replied:
Handling the case where <3 x i32> is widening makes sense to me, but I think I'm missing something with the `SrcEltSize == LegalEltSize` check. When will a vector's element type get legalized to different size? Won't the two always be the same?

preames replied:
When writing an updated description for this review which was inclusive of the splitting behavior in the cost model, I sat down to write a couple of lowering tests to be sure the costing I was describing was both correct and matched actual lowering.

When doing so, I found a test case which crashes during lowering (see below).  Given that, I added a restriction to this change to specifically only change the cost when promoting.  This still covers the cases I'm interesting in, and avoids the possibility of perturbing the costs in an area where we appear to have a lowering problem.

```
; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
; RUN: llc -mtriple=riscv64 -mattr=+zve32x < %s | FileCheck %s

; Note that vscale x 32 x i32 appears to work just fine, and we split to two LMUL 8
; types as expected.  It's apparently splitting combined with the need to lower the
; second half with VL predication which is causing problems.
define void @test31(ptr %a, ptr %b) nounwind vscale_range(4,1024) {
  %LD = call <vscale x 31 x i32> @llvm.masked.load.nxv31i32.p0(ptr %a, i32 4, <vscale x 31 x i1> splat(i1 true), <vscale x 31 x i32> poison)
  store <vscale x 31 x i32> %LD, ptr %b
  ret void
}




```

preames replied:
> Handling the case where <3 x i32> is widened makes sense to me, but I think I'm missing something with the `SrcEltSize == LegalEltSize` check. When will a vector's element type get legalized to different size? Won't the two always be the same?

I'd originally asserted this locally, and the assert failed fairly widely.  I didn't dig into it in any detail, but I think it involved a vector of i1.

lukel97 replied:
I checked out this branch locally and added in the assert, it didn't seem to trigger. Maybe something changed in the meantime? We have legal i1 vector types so I presume if it was to get legalized the element type would still be i1.

lukel97 replied:
For the record I'm happy to leave this in defensively. Everything else LGTM

preames replied:
Thanks for pushing on this.  I went back and ran the local tests with the assert, and found out that my assert/check was actually entirely spurious.  

The case that I found locally (without the isKnownLE check) is:
```
<2 x i1> 
v2i1 
```

After staring at that for a moment, and going "huh?", I realized that the check was using the Store size of the element from DL, and the non-store width of the element from the VT.  Fixing that mismatch causes the assert to never fail.  

Given that, I'm just going to remove the check entirely.

preames committed:
Restrict to promotion case only

The logic should be correct for splitting as well, but when testing out the
actual codegen for splitting, I found a case where we just crash.  Given
that, let's avoid perturbing costs for cases which might need split for
now.

preames commented on llvm/test/Transforms/LoopVectorize/RISCV/short-trip-count.ll:
```diff
@@ -7,24 +7,22 @@ define void @small_trip_count_min_vlen_128(ptr nocapture %a) nounwind vscale_ran
 ; CHECK-NEXT:    br label [[VECTOR_BODY:%.*]]
 ; CHECK:       vector.body:
 ; CHECK-NEXT:    [[INDEX:%.*]] = phi i32 [ 0, [[VECTOR_PH]] ], [ [[INDEX_NEXT:%.*]], [[VECTOR_BODY]] ]
-; CHECK-NEXT:    [[TMP7:%.*]] = add i32 [[INDEX]], 0
```
For this test, key details are:
<vscale x 2 x i32> is legal
<vscale x 1 x i32> is not
scalarization (the prior costing approach) returns Invalid for scalable vectors
our actual lowering uses the VL predicated <vscale x 2 x i32>

So this is a valid and correct change.  I make no comment on whether the vectorizer chose the cheapest option.  :)

topperc commented on llvm/test/Transforms/LoopVectorize/RISCV/short-trip-count.ll:
```diff
@@ -67,23 +65,23 @@ define void @small_trip_count_min_vlen_32(ptr nocapture %a) nounwind vscale_rang
 ; CHECK-NEXT:    [[TMP1:%.*]] = mul i32 [[TMP0]], 4
-; CHECK-NEXT:    [[TMP4:%.*]] = sub i32 [[TMP1]], 1
-; CHECK-NEXT:    [[N_RND_UP:%.*]] = add i32 4, [[TMP4]]
+; CHECK-NEXT:    [[TMP2:%.*]] = sub i32 [[TMP1]], 1
```
Is the script just doing renames here?

preames committed:
Merge branch 'main' into pr-riscv-oddvec-ldst-cost

preames committed:
Remove pointless (and incorrect) check

The check I'd written used Store size on one side, and non-store size on the
other.  For e.g. i1, that always failed.

preames committed:
Merge branch 'main' into pr-riscv-oddvec-ldst-cost
